packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
   } while (last);
   return next;
 }
+
+
+class Tree {
+  private root: TreeNode;
+  private fastMap = new Map<ScopeRef, TreeNode>();
```

Comment:
You could probably get rid of this map if you returned nodes from `addTreeNode` and passed those around (e.g. through context, to hooks, etc.) rather than only the ref object. Then you wouldn't need to look up nodes from scopes, you'd have the nodes which contain the scope instead.

Reply:
hmmm that's a larger change, that'd mean changing activeScope to a TreeNode and passing a TreeNode through context in place of parent
I'm fine doing that but it might be harder to review? Maybe followup?

Reply:
sure, might be good followup. I think it could potentially clean up the code a little